### ISSUE TITLE
pass context throughout

### ISSUE THREAD

**Issue by jacobweinstock** [stmcginnis/gofish#104]
Hey, great library! thanks so much for all the hard work!

I was wondering if you'd be open to a PR to add context throughout? It'd obviously be a pretty breaking change as most of the function/method signatures would need to be changed.

**Comment by stmcginnis** [comment]
Thanks Jacob. Could you explain a little more what you are trying to do?

**Comment by jacobweinstock** [comment]
Ah sorry, yes. I'm interested in being able to pass in a context and have `gofish` honor the timeout, deadline, or cancelation.  

Right now, I know I can set a timeout by using the timeout in the `http.Client` struct and passing that into `Connect` or using the `ClientConfig`.  My challenge is that when importing `gofish` and using it with other libraries they can pass around a context and use/understand the timeout/deadline/cancel functionality. 

With current `gofish` I need to translate the context timeout/deadline to a number and then use that in `gofish`. I also haven't found a clean way to honor a canceled context. 

Hope that makes sense.

**Comment by stmcginnis** [comment]
I do think that would end up being too disruptive.

Really, the client object is the context that already gets passed around. Can we add something there that would help you?

**Comment by jacobweinstock** [comment]
Sure, totally understand. The challenge with the client object is that putting `context.Context` into a struct is not ideal or recommended. 
> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it.
https://golang.org/pkg/context/

Thanks for the conversation and the consideration though. I will have to rethink on my end how to proceed.

**Comment by vinialves** [comment]
Hello guys, good discussion here!

Connect receives the struct ClientConfig, and as explained the recommended golang way to send the context is sending it as the first parameter in the function args.

To keep Connect working for previous Gofish versions and to avoid sending Context in the ClientConfig struct (that is not the prefered idiom to send the context as explained) we could create a func ConnectContext(ctx context.Context, config ClientConfig) (c *APIClient, err error)

We could change Connect and ConnectDefault to get the context from context.Background()

In order to implement as recommended by golang we would need to do changes in all functions to send the context as first parameter to "runRequest" in "client.go" where we would replace the NewRequest call:
req, err := http.NewRequest(method, endpoint, payloadBuffer)
req, err := http.NewRequestWithContext(ctx, endpoint, payloadBuffer)

I think it would be too many work just to follow the recommendation. In my opinion, we could set the context in the APIClient struct then we get the context from the struct during runRequest. It is not the preferred idiom to send the context through the functions, but it is still ok in my opinion.

I think this lib is doing something similar: https://github.com/spf13/cobra/blob/master/command.go
See ExecuteContext: https://github.com/spf13/cobra/blob/master/command.go#L886
Also, see where it gets default context: https://github.com/spf13/cobra/blob/master/command.go#L902

**Comment by stmcginnis** [comment]
This seems reasonable to me. Thanks for the good idea!

I'm pretty tied up with other things right now, so I won't be able to get to this myself for awhile. But if someone wants to put up a PR, I'd be happy to get that reviewed as soon as I can.

**Comment by vinialves** [comment]
thanks @stmcginnis maybe I can do this PR in the future with other changes I am planning to work. I just need to put it in the queue as I also plan to create other PRs for Gofish like the ability to get redfish registries (which contains the messages for the redfish events). Also, I saw  It was added Context (not the same context we are discussing here) and Protocol in the creation of the event subscriptions (sorry I forgot to add these attributes in the PR previously), good job! I am already doing changes to get last Gofish version and start creating subscriptions with context :)
